tools/pythonpkg/tests/fast/pandas/test_df_object_resolution.py

```diff
@@ -0,0 +1,370 @@
+
+    def test_list_value_upgrade(self, duckdb_cursor):
+        x = pd.DataFrame(
+            [
+                {'0': [['5'], [34], [-245]]}
+            ]
+        )
+        duckdb_col = duckdb.query("select [['5'], ['34'], ['-245']] as '0'").df()
+        converted_col = duckdb.query_df(x, "tbl", "select * from tbl").df()
+        pd.testing.assert_frame_equal(duckdb_col, converted_col)
+
+    def test_list_column_value_upgrade(self, duckdb_cursor):
+        x = pd.DataFrame(
+            [
+                [ [1, 25, 300] ],
+                [ [500, 345, 30] ],
+                [ [50, 'a', 67] ],
+            ]
+        )
+        x.rename(columns = {0 : 'a'}, inplace = True)
+        converted_col = duckdb.query_df(x, "x", "select * from x").df()
+        duckdb.query("""
+            CREATE TABLE tmp3(
+                a VARCHAR[]
+            );
+        """)
+        duckdb.query("""
+            INSERT INTO tmp3 VALUES (['1', '25', '300'])
+        """)
+        duckdb.query("""
+            INSERT INTO tmp3 VALUES (['500', '345', '30'])
+        """)
+        duckdb.query("""
+            INSERT INTO tmp3 VALUES (['50', 'a', '67'])
+        """)
+        duckdb_col = duckdb.query("select a from tmp3 AS '0'").df()
+        print(duckdb_col.columns)
+        print(converted_col.columns)
+        pd.testing.assert_frame_equal(converted_col, duckdb_col)
+
+    def test_ubigint_object_conversion(self, duckdb_cursor):
+        data = [18446744073709551615, 0]
+        x = pd.DataFrame({'0': pd.Series(data=data, dtype='object')})
+        converted_col = duckdb.query_df(x, "x", "select * from x").df()
+        uint64_dtype = np.dtype('uint64')
+        assert isinstance(converted_col['0'].dtype, uint64_dtype.__class__) == True
+
+    def test_double_object_conversion(self, duckdb_cursor):
+        data = [18446744073709551616, 0]
+        x = pd.DataFrame({'0': pd.Series(data=data, dtype='object')})
+        converted_col = duckdb.query_df(x, "x", "select * from x").df()
+        double_dtype = np.dtype('float64')
+        assert isinstance(converted_col['0'].dtype, double_dtype.__class__) == True
+
+    def test_integer_conversion_fail(self, duckdb_cursor):
+        data = [2**10000, 0]
+        x = pd.DataFrame({'0': pd.Series(data=data, dtype='object')})
+        converted_col = duckdb.query_df(x, "x", "select * from x").df()
+        print(converted_col['0'])
+        double_dtype = np.dtype('object')
+        assert isinstance(converted_col['0'].dtype, double_dtype.__class__) == True
+
+    def test_fallthrough_object_conversion(self, duckdb_cursor):
+        x = pd.DataFrame(
+            [
+                [IntString(4)],
+                [IntString(2)],
+                [IntString(0)],
+            ]
+        )
+        duckdb_col = duckdb.query_df(x, "x", "select * from x").df()
+        df_expected_res = pd.DataFrame({'0': pd.Series(['4','2','0'])})
+        pd.testing.assert_frame_equal(duckdb_col, df_expected_res)
```

Comment:
Could we add tests with:

*A large data frame where the sampling picks up the incorrect type (e.g. we have 10000 times the element `1` and 1 time the element `[1, 2, 3]`)
* Many mixed objects (e.g. `1, [1], {'a': 42}, 1.0`)
* Mixed numeric objects (e.g. `[1, None, 1.0, decimal.Decimal(1.0)]`)
* More recursive nested types (e.g. lists of structs, structs containing lists, including nulls at various levels), e.g.:


```py
# lists of lists
numpy.array([
	[[1, 2, 3], None, [None, 4, 5]],
	None,
	[[6, None, 8], [9], [], None]
], dtype=object)

# structs with lists
numpy.array([
	{'a': [1, None, 3], 'b': 42},
	None,
	{'a': None, 'b': None},
	{'a': [4, None, 6], 'b': None}
], dtype=object)
```

Reply:
py::handle only overloads [] for `const char*` and `py::handle` so the closest I can get it is this:
`auto next_item_type = GetItemType(column[py::int_(i)], can_convert);`

Which I'm not super excited about

Reply:
That solution looks good to me.

Reply:
I will add more tests with recursive nested types, the only problem I have there is creating the reference tables. I wrote something (pretty horrible) for this when I was writing tests for testing arrow list -> duckdb, only problem is it uses TransformPythonValue (for the (?) resolution) and that's also what is used to transform the values of the panda dataframe columns.

```py
def create_and_register_comparison_result(column_list, duckdb_conn):
    columns = ",".join([f'{name} {dtype}' for (name, dtype, _) in column_list])
    column_amount = len(column_list)
    assert(column_amount)
    row_amount = len(column_list[0][2])
    inserted_values = []
    for row in range(row_amount):
        for col in range(column_amount):
            inserted_values.append(transform(column_list[col][2][row]))
    inserted_values = tuple(inserted_values)

    column_format = ",".join(['?' for _ in range(column_amount)])
    row_format = ",".join([f"({column_format})" for _ in range(row_amount)])
    query = f"""CREATE TABLE test ({columns});
        INSERT INTO test VALUES {row_format};
    """
    duckdb_conn.execute(query, inserted_values)
  ```

Reply:
How about just using SQL:

create table tbl as select [[1,2,3],[4,5]] col union all select null;